habitat_code/ramachandran_code_for_review.py

```diff
@@ -0,0 +1,511 @@
+
+# B06011_001E: Estimate!!Median income in the past 12 months --!!Total:	
+# Sources: https://api.census.gov/data/2019/acs/acs5/variables.html; https://pypi.org/project/census/
+
+# Obtain Census variables from the 2019 American Community Survey at the tract level for Illinois (FIPS code: 17)
+census_path = os.path.join(data_dir, 'census_data_for_colorado.csv')
+
+if os.path.exists(census_path):
+    census_df = pd.read_csv(census_path)
+    print("Data is already downloaded.")
+
+else:
+    # Authenticate
+    api_key = getpass.getpass('U.S. Census API Key')
+    c = Census(api_key)
+    c
+
+    CO_census = c.acs5.state_county_tract(fields = ('NAME', 'B06011_001E'),
+                                      state_fips = '08',
+                                      county_fips = "*",
+                                      tract = "*",
+                                      year = 2021)
+    
+    tracts_census_df = (pd.DataFrame(CO_census, 
+                                     columns=['NAME', 
+                                              'B06011_001E', 
+                                              'state', 
+                                              'county', 
+                                              'tract'])
+                        .to_csv(census_path, index=False))
+    
+    census_df = pd.read_csv(census_path)
+
+
+census_df = census_df.rename(columns={'B06011_001E': 'median_income'})
+census_df.head()
+
+# Merge census data with tracts
+
+boulder_city_tracts_gdf
+
+# %%
+# Merge datasets
+
+boulder_city_tracts_gdf = boulder_city_tracts_gdf.loc[:, ['TRACTCE', 'geometry', 'NAMELSAD']]
+
+boulder_city_tracts_gdf['TRACTCE'] = pd.to_numeric(boulder_city_tracts_gdf['TRACTCE'], errors='coerce')
+
+tracts_w_census_gdf = boulder_city_tracts_gdf.merge(census_df, left_on='TRACTCE', right_on='tract')
+
+negative_or_not = (tracts_w_census_gdf['median_income'] >= 0)
+
+tracts_w_census_gdf = tracts_w_census_gdf[negative_or_not]
+tracts_w_census_gdf.head()
+
+# %%
+tracts_w_census_gdf['median_income'].plot.hist()
+
+# %%
+tracts_w_census_gdf.plot('median_income', legend=True)
+
+# %%
+# Test whether tracts are correctly georeferenced
+
+tracts_basemap_plot = tracts_w_census_gdf.hvplot(geo=True, alpha=0.3, tiles='EsriImagery')
+tracts_basemap_plot
+
+# %%
+# Download data using Microsoft Planetary Computer STAC catalog
+
+# Access catalog
+pc_catalog = pystac_client.Client.open(
+    "https://planetarycomputer.microsoft.com/api/stac/v1"
+)
+
+pc_catalog.title
+
+# %%
+# Check if any data has already been downloaded and processed
+
+all_greenspace_stats_path = os.path.join(data_dir, 'all_greenspace_stats.csv')
+
+if os.path.exists(all_greenspace_stats_path):
+    print("All greenspace stats file exists. Checking "
+          "which tracts have been downloaded and processed...")
+
+    # Extract tracts to be calculated
+    all_tracts = tracts_w_census_gdf['TRACTCE'].astype('int64')
+
+    # Load tracts already processed
```

Comment:
I like this extra caching step to scrape tracts that have already been processed, in case execution is interrupted. Maybe there is some potential to streamline this process with a method?